sdks/go/pkg/beam/core/runtime/exec/fn.go

```diff
@@ -125,10 +125,148 @@ func Invoke(ctx context.Context, ws []typex.Window, ts typex.EventTime, fn *func
 	return nil, nil
 }
 
+// InvokeWithoutEventTime runs the given function at time 0 in the global window.
 func InvokeWithoutEventTime(ctx context.Context, fn *funcx.Fn, opt *MainInput, extra ...interface{}) (*FullValue, error) {
 	return Invoke(ctx, window.SingleGlobalWindow, mtime.ZeroTimestamp, fn, opt, extra...)
 }
 
+// invoker is a container class for hot path invocations of DoFns, to avoid
+// repeating fixed set up per element.
+type invoker struct {
+	fn   *funcx.Fn
+	args []interface{}
+	// replace with a slice of functions to run over the args slice? Profile and find out!
+	ctxIdx, wndIdx, etIdx int   // specialized input indexes
+	outEtIdx, errIdx      int   // specialized output indexes
+	in, out               []int // general indexes
+}
+
+func newInvoker(fn *funcx.Fn) *invoker {
+	n := &invoker{
+		fn:       fn,
+		args:     make([]interface{}, len(fn.Param)),
+		ctxIdx:   -1,
+		wndIdx:   -1,
+		etIdx:    -1,
+		outEtIdx: -1,
+		errIdx:   -1,
+		in:       fn.Params(funcx.FnValue | funcx.FnIter | funcx.FnReIter | funcx.FnEmit),
+		out:      fn.Returns(funcx.RetValue),
+	}
+	if index, ok := fn.Context(); ok {
+		n.ctxIdx = index
+	}
+	if index, ok := fn.Window(); ok {
+		n.wndIdx = index
+	}
+	if index, ok := fn.EventTime(); ok {
+		n.etIdx = index
+	}
+	if index, ok := fn.OutEventTime(); ok {
+		n.outEtIdx = index
+	}
+	if index, ok := fn.Error(); ok {
+		n.errIdx = index
+	}
+	return n
+}
+
+// ClearArgs zeroes argument entries in the cached slice to allow values to be garbage collected after the bundle ends.
+func (n *invoker) ClearArgs() {
+	for i := range n.args {
```

Comment:
That alloc is part of what we're trying to avoid, though it's probably fine on a per-bundle level.
Both approaches need to zero the memory, this way is less work for the CPU compared to getting something new from the heap.